File: backend/app/schemas/knowledge.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict


class ORMModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class KnowledgeNodeBase(ORMModel):
    id: int
    slug: str
    title: str
    layer: int
    category: str
    difficulty: str
    summary: str
    tags: list[str]
    prerequisites: list[str]
    next_topics: list[str]
    related_topics: list[str]


class KnowledgeNodeRead(KnowledgeNodeBase):
    definition: str
    intuition: str
    why_it_matters: str
    math_form: str
    formulas: list[str]
    code_example: str
    applications: list[str]
    misconceptions: list[str]
    prerequisites: list[str]
    next_topics: list[str]
    related_topics: list[str]
    recommended_resources: list[str]
    self_check_questions: list[str]
    extension_questions: list[str]
    created_at: datetime
    updated_at: datetime


class KnowledgeListResponse(BaseModel):
    items: list[KnowledgeNodeBase]
    total: int


class KnowledgeLayer(BaseModel):
    layer: int
    name: str
    count: int


class RelatedKnowledgeResponse(BaseModel):
    slug: str
    relation_type: Literal["prerequisite", "next", "related", "same_layer"]
    items: list[KnowledgeNodeBase]


class GraphNode(BaseModel):
    id: str
    title: str
    layer: int
    category: str
    difficulty: str
    tags: list[str]


class GraphEdge(BaseModel):
    source: str
    target: str
    relation_type: str


class GraphResponse(BaseModel):
    nodes: list[GraphNode]
    edges: list[GraphEdge]


class SearchResult(KnowledgeNodeBase):
    highlights: list[str] = []

File: frontend/src/types/knowledge.ts
export interface KnowledgeSummary {
  id: number;
  slug: string;
  title: string;
  layer: number;
  category: string;
  difficulty: string;
  summary: string;
  tags: string[];
  prerequisites: string[];
  next_topics: string[];
  related_topics: string[];
}

export interface KnowledgeNode extends KnowledgeSummary {
  definition: string;
  intuition: string;
  why_it_matters: string;
  math_form: string;
  formulas: string[];
  code_example: string;
  applications: string[];
  misconceptions: string[];
  prerequisites: string[];
  next_topics: string[];
  related_topics: string[];
  recommended_resources: string[];
  self_check_questions: string[];
  extension_questions: string[];
}

export interface KnowledgeList {
  items: KnowledgeSummary[];
  total: number;
}

export interface KnowledgeResource {
  id: number;
  knowledge_slug: string;
  title: string;
  url: string;
  source: string;
  resource_type: string;
  authority_level: "S" | "A" | "B" | "C" | string;
  difficulty: "beginner" | "intermediate" | "advanced" | string;
  language: "en" | "zh" | "other" | string;
  estimated_time: string;
  description: string;
  why_recommended: string;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface ResourceList {
  items: KnowledgeResource[];
  total: number;
}

export interface KnowledgeLayer {
  layer: number;
  name: string;
  count: number;
}

export interface GraphNode {
  id: string;
  title: string;
  layer: number;
  category: string;
  difficulty: string;
  tags: string[];
}

export interface GraphEdge {
  source: string;
  target: string;
  relation_type: string;
}

export interface GraphData {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface Settings {
  id: number;
  user_name: string;
  preferred_style: string;
  llm_provider: string;
  api_base_url: string;
  ai_enabled: boolean;
  max_rag_chunks: number;
}
